main/lsp/LSPIndexer.cc

```diff
@@ -144,26 +90,75 @@ vector<core::FileHash> LSPIndexer::computeFileHashes(const vector<shared_ptr<cor
     });
 
     {
-        vector<pair<int, core::FileHash>> threadResult;
+        vector<pair<int, unique_ptr<core::FileHash>>> threadResult;
         for (auto result = resultq->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), logger); !result.done();
              result = resultq->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), logger)) {
             if (result.gotItem()) {
                 for (auto &a : threadResult) {
-                    res[a.first] = move(a.second);
+                    files[a.first]->setFileHash(move(a.second));
                 }
             }
         }
     }
-    return res;
 }
 
-vector<core::FileHash> LSPIndexer::computeFileHashes(const vector<shared_ptr<core::File>> &files) const {
-    return computeFileHashes(files, *emptyWorkers);
+void LSPIndexer::computeFileHashes(const vector<shared_ptr<core::File>> &files) const {
+    computeFileHashes(files, *emptyWorkers);
 }
 
-bool LSPIndexer::canTakeFastPath(const SorbetWorkspaceEditParams &params,
-                                 const vector<core::FileHash> &fileHashes) const {
-    return canTakeFastPathInternal(*initialGS, *config, globalStateHashes, fileHashes, params.updates);
+bool LSPIndexer::canTakeFastPath(const std::vector<std::shared_ptr<core::File>> &changedFiles,
+                                 bool containsPendingTypecheckUpdates) const {
+    Timer timeit(config->logger, "fast_path_decision");
+    auto &logger = *config->logger;
+    logger.debug("Trying to see if fast path is available after {} file changes", changedFiles.size());
+    if (config->disableFastPath) {
+        logger.debug("Taking slow path because fast path is disabled.");
+        prodCategoryCounterInc("lsp.slow_path_reason", "fast_path_disabled");
+        return false;
+    }
+
+    const UnorderedMap<int, shared_ptr<core::File>> emptyMap;
+    const UnorderedMap<int, shared_ptr<core::File>> &evictedFilesRef =
+        containsPendingTypecheckUpdates ? evictedFiles : emptyMap;
+    for (auto &f : changedFiles) {
+        auto fref = initialGS->findFileByPath(f->path());
+        if (!fref.exists()) {
+            logger.debug("Taking slow path because {} is a new file", f->path());
+            prodCategoryCounterInc("lsp.slow_path_reason", "new_file");
+            return false;
+        } else {
+            const auto &oldFile = getOldFile(fref, *initialGS, evictedFilesRef);
```

Comment:
This is where we consume `evictedFiles` (or the empty map if the files don't contain the pending update).